src/lava/proc/sparse/models.py

```diff
@@ -0,0 +1,359 @@
+        # accumulation at timestep t-1. This prevents deadlocking in
+        # networks with recurrent connectivity structures.
+        self.a_out.send(self.a_buff)
+        if self.num_message_bits.item() > 0:
+            s_in = self.s_in.recv()
+            self.a_buff = self.weights.dot(s_in)
+        else:
+            s_in = self.s_in.recv().astype(bool)
+            self.a_buff = self.weights[:, s_in].sum(axis=1).A1
+
+        self.recv_traces(s_in)
+
+
+@implements(proc=LearningSparse, protocol=LoihiProtocol)
+@requires(CPU)
+@tag("bit_approximate_loihi", "fixed_pt")
+class PyLearningSparseModelBitApproximate(
+        LearningConnectionModelBitApproximate, AbstractPySparseModelBitAcc):
+    """Implementation of Conn Process with Sparse synaptic connections that is
+    uses similar constraints as the  Loihi's hardware implementation of Sparse,
+    but does not mimics Loihi behaviour bit-by-bit.
+
+    Warning: LearningSparse on CPU is not offereing any memory usage benefits
+    over using LearningDense.
+    """
+    # overwrite dense PyTypes with sparse
+    tag_1: csr_matrix = LavaPyType(csr_matrix, int, precision=8)
+    tag_2: csr_matrix = LavaPyType(csr_matrix, int, precision=6)
+
+    def __init__(self, proc_params):
+        super().__init__(proc_params)
+        # Flag to determine whether weights have already been scaled.
+        self.num_weight_bits: int = self.proc_params.get("num_weight_bits", 8)
+
+    def run_spk(self):
+        self.weight_exp: int = self.proc_params.get("weight_exp", 0)
+
+        # Since this Process has no learning, weights are assumed to be static
+        # and only require scaling on the first timestep of run_spk().
+        if not self.weights_set:
+            self.weights = truncate_weights(
+                self.weights,
+                sign_mode=self.sign_mode,
+                num_weight_bits=self.num_weight_bits
+            )
+            self.weights_set = True
+
+        # The a_out sent at each timestep is a buffered value from dendritic
+        # accumulation at timestep t-1. This prevents deadlocking in
+        # networks with recurrent connectivity structures.
+        self.a_out.send(self.a_buff)
+        if self.num_message_bits.item() > 0:
+            s_in = self.s_in.recv()
+            a_accum = self.weights.dot(s_in)
+        else:
+            s_in = self.s_in.recv().astype(bool)
+            a_accum = self.weights[:, s_in].sum(axis=1).A1
+
+        self.a_buff = (
+            np.left_shift(a_accum, self.weight_exp)
+            if self.weight_exp > 0
+            else np.right_shift(a_accum, -self.weight_exp)
+        )
+
+        self.recv_traces(s_in)
+
+
+class AbstractPyDelaySparseModel(PyLoihiProcessModel):
+    """Abstract Conn Process with Sparse synaptic connections which incorporates
+    delays into the Conn Process.
+    """
+
+    @staticmethod
+    def get_del_wgts(weights, delays) -> spmatrix:
```

Comment:
I have trouble understanding this; both the docstring and the code. Can this be clarified?

Reply:
see comment above.